### PR TITLE
Clustering updates

### PR DESCRIPTION
**Pull request by derekcollison**
Added in support for stream update as well as stream and consumer detailed lists which use a scatter and gather approach.

/cc @nats-io/core

### CODE REVIEW

**Comment by derekcollison** [comment]
I had these alot more stable but now seem to be flapping again. Taking a short break and then will look into getting this stable again.

File: server/consumer.go [hunk 1]
```diff
@@ -228,13 +230,13 @@ func (mset *Stream) AddConsumer(config *ConsumerConfig) (*Consumer, error) {
...
 	mset.mu.RLock()
-	jsa := mset.jsa
+	s, jsa := mset.srv, mset.jsa
 	mset.mu.RUnlock()
 
 	// If we do not have the consumer assigned to us in cluster mode we can not proceed.
 	// Running in single server mode this always returns true.
 	if oname != _EMPTY_ && !jsa.consumerAssigned(mset.Name(), oname) {
-		return nil, ErrJetStreamNotAssigned
+		s.Debugf("Consumer %q > %q does not seem to be assigned to this server", mset.Name(), oname)
```
Comment (kozlovic) [inline server/consumer.go:239]: So now we are not returning with an error like we used to. Is that intentional? The comment above seem to indicate that if not assigned we cannot proceed, yet we do.

File: server/consumer.go [hunk 2]
```diff
@@ -603,6 +608,16 @@ func (o *Consumer) setLeader(isLeader bool) {
 		// Restore our saved state. During non-leader status we just update our underlying store.
 		o.readStoredState()
 
+		// Do info sub.
+		if o.infoSub == nil && mset != nil && mset.jsa != nil {
+			mset.mu.RLock()
```
Comment (kozlovic) [inline server/consumer.go:613]: I don't know enough of the locking rules between stream and consumer, but here it seem we have o.mu.Lock() above and now trying to get mset's lock. Could that cause a lock inversion (again without knowing, I would think that it seem more logical to have stream -> consumer than the opposite).

(if that is an issue, would move those mset's fields capture before getting o.mu.Lock(), even if we end-up not needing them).

File: server/jetstream_api.go [hunk 3]
```diff
@@ -1024,6 +1026,12 @@ func (s *Server) jsStreamListRequest(sub *subscription, c *client, subject, repl
 		offset = req.Offset
 	}
 
+	// Clustered mode will invoke a scatter and gather.
+	if s.JetStreamIsClustered() {
+		s.startGoRoutine(func() { s.jsClusteredStreamListRequest(acc, ci, offset, subject, reply, msg) })
```
Comment (kozlovic) [inline server/jetstream_api.go:1031]: Don't we need to copy some of the content of those variables here? Meaning when the stack unrolls, those fields may be overwritten (say msg points to wire, etc..).

File: server/jetstream_api.go [hunk 4]
```diff
@@ -1845,6 +1855,13 @@ func (s *Server) jsConsumerListRequest(sub *subscription, c *client, subject, re
 	}
 
 	streamName := streamNameFromSubject(subject)
+
+	// Clustered mode will invoke a scatter and gather.
+	if s.JetStreamIsClustered() {
+		s.startGoRoutine(func() { s.jsClusteredConsumerListRequest(acc, ci, offset, streamName, subject, reply, msg) })
```
Comment (kozlovic) [inline server/jetstream_api.go:1861]: Same here.

File: server/jetstream_cluster.go [hunk 5]
```diff
@@ -1001,10 +1013,10 @@ func (js *jetStream) applyStreamEntries(mset *Stream, ce *CommittedEntry) (bool,
 				if err != nil {
 					panic(err.Error())
 				}
-				if lseq == 0 && mset.lastSeq() != 0 { // Very first msg
+				// Skip by hand here since first msg special case.
```
Comment (kozlovic) [inline server/jetstream_cluster.go:1016]: Not sure I fully understand this comment :-)

File: server/jetstream_cluster.go [hunk 6]
```diff
@@ -1136,7 +1148,10 @@ func (js *jetStream) processStreamAssignment(sa *streamAssignment) {
 	accStreams[stream] = sa
 	cc.streams[acc.Name] = accStreams
 
-	isMember := sa.Group.isMember(cc.meta.ID())
+	var isMember bool
+	if sa.Group != nil && cc.meta != nil {
```
Comment (kozlovic) [inline server/jetstream_cluster.go:1152]: If those are possible legitimate conditions (group and/or meta being nil), a small comment indicating what that means may help the reader.

File: server/jetstream_cluster.go [hunk 7]
```diff
@@ -1918,6 +1936,207 @@ func (s *Server) jsClusteredStreamPurgeRequest(ci *ClientInfo, stream, subject,
...
+		case rc <- &si:
+		default:
+			s.Warnf("Failed placing stream info result on internal chan")
+		}
+	})
+	defer s.sysUnsubscribe(rsub)
+
+	// Send out our requests here.
+	for _, sa := range streams {
+		isubj := fmt.Sprintf("$JSC.SI.%s.%s", sa.Client.Account, sa.Config.Name)
```
Comment (kozlovic) [inline server/jetstream_cluster.go:1994]: Should "$JSC.SI.." made it a constant (like the template that we have)?

File: server/jetstream_cluster.go [hunk 8]
```diff
@@ -1918,6 +1936,207 @@ func (s *Server) jsClusteredStreamPurgeRequest(ci *ClientInfo, stream, subject,
...
+		case rc <- &ci:
+		default:
+			s.Warnf("Failed placing consumer info result on internal chan")
+		}
+	})
+	defer s.sysUnsubscribe(rsub)
+
+	// Send out our requests here.
+	for _, ca := range consumers {
+		isubj := fmt.Sprintf("$JSC.CI.%s.%s.%s", ca.Client.Account, stream, ca.Name)
```
Comment (kozlovic) [inline server/jetstream_cluster.go:2097]: Same here.

File: server/stream.go [hunk 9]
```diff
@@ -153,7 +155,7 @@ func (a *Account) addStream(config *StreamConfig, fsConfig *FileStoreConfig, sa
 	// If we do not have the stream assigned to us in cluster mode we can not proceed.
 	// Running in single server mode this always returns true.
 	if !jsa.streamAssigned(config.Name) {
-		return nil, ErrJetStreamNotAssigned
+		s.Debugf("Stream %q does not seem to be assigned to this server", config.Name)
```
Comment (kozlovic) [inline server/stream.go:158]: Again, here the comment says that we cannot proceed but we do and just log a debug message. If it's ok, let fix the comment.

File: server/stream.go [hunk 10]
```diff
@@ -297,14 +302,25 @@ func (mset *Stream) setLeader(isLeader bool) error {
 }
 
 // Lock should be held.
-func (mset *Stream) startSyncSub() {
+func (mset *Stream) startClusterSubs() {
+	if mset.infoSub == nil {
+		if jsa := mset.jsa; jsa != nil {
+			isubj := fmt.Sprintf("$JSC.SI.%s.%s", jsa.acc(), mset.config.Name)
```
Comment (kozlovic) [inline server/stream.go:308]: If we make this a constant, replace the string here.

File: server/consumer.go (hunk 1 above)
Comment (derekcollison) [reply server/consumer.go:239]: This is all for raciness during startup. We should warn but could be that we process a recovered stream or consumer before we process the assignment if it was the last. In normal practice does not happen and if it is truly not supposed to be there the deleteOp will follow from meta and delete it anyway.

File: server/consumer.go (hunk 2 above)
Comment (derekcollison) [reply server/consumer.go:613]: Good point, will cleanup.

File: server/jetstream_api.go (hunk 4 above)
Comment (derekcollison) [reply server/jetstream_api.go:1861]: Apologies, same what?

File: server/jetstream_cluster.go (hunk 5 above)
Comment (derekcollison) [reply server/jetstream_cluster.go:1016]: ok will make clearer. sequences are unsigned, and each message we replicate has the prior sequence it expects when being placed in the stream storage. So 0 can't have prev of -1.

File: server/jetstream_cluster.go (hunk 6 above)
Comment (derekcollison) [reply server/jetstream_cluster.go:1152]: It should not happen, if it does, at least for now, that is an error.

File: server/jetstream_cluster.go (hunk 7 above)
Comment (derekcollison) [reply server/jetstream_cluster.go:1994]: Yes will do that.

File: server/jetstream_cluster.go (hunk 8 above)
Comment (derekcollison) [reply server/jetstream_cluster.go:2097]: Yep.

File: server/stream.go (hunk 9 above)
Comment (derekcollison) [reply server/stream.go:158]: Yes will correct.

File: server/stream.go (hunk 10 above)
Comment (derekcollison) [reply server/stream.go:308]: Done

File: server/jetstream_api.go (hunk 4 above)
Comment (kozlovic) [reply server/jetstream_api.go:1861]: Same than this comment I made that you missed: https://github.com/nats-io/nats-server/pull/1819#discussion_r559226595

File: server/jetstream_api.go (hunk 3 above)
Comment (derekcollison) [reply server/jetstream_api.go:1031]: Ah I see what you mean. Will take a look more closely. Thanks.

File: server/events.go [hunk 11]
```diff
@@ -1673,7 +1673,9 @@ func (s *Server) debugSubscribers(sub *subscription, c *client, subject, reply s
 		}
 		// Cleanup the WC entry.
 		s.mu.Lock()
-		delete(s.sys.replies, replySubj)
+		if s.sys != nil && s.sys.replies != nil {
+			delete(s.sys.replies, replySubj)
+		}
```
Comment (kozlovic) [inline server/events.go:1678]: Not sure it makes sense to send the reply if we did not delete from the map, but I guess it does not hurt..

Comment (derekcollison) [reply server/events.go:1678]: True will fix..

**Review by kozlovic — APPROVED** [review]
LGTM

**Pull request merged**